Support audio

strvert opened the pull request:
I understand that Screego aims to remain extremely simple, and that audio support was previously closed in issue #102. However, in my daily work, I find it inconvenient not to be able to share audio when I need to demonstrate code and its results to colleagues. As a game programmer, a lot of sound is produced as a result of my code. Not being able to convey this to my colleagues is a significant drawback.

Considering the context of issue #102, I have implemented a slightly different concept, which I will explain below.

## Audio Transmission Settings
![image](https://github.com/screego/server/assets/20394825/2d54d312-486b-4f4e-9825-a708d6a0aa64)

In the browser menu, users can choose whether to share audio. If enabled, both video and audio tracks will be sent; if disabled, only the video track will be sent. In browsers that do not support audio sharing, such as Firefox, this menu will not be displayed, and only the video track will always be sent.

## Audio Listening Settings
First, audio can only be listened to under the following conditions:

1. It is not the hostStream.
2. It is the selectedStream.
3. The "Hear Audio" button is enabled.

The reason for 1 is straightforward—nobody wants to hear their own screen's audio twice.

For 2, the concept of this implementation is the reason. When receiving multiple screens, I couldn't think of a useful scenario where overlapping audio is helpful. Also, having many mute buttons displayed would not be simple. Therefore, by limiting the audio track used for playback to the selected stream, users can choose which audio to hear.

For 3, this is the only additional button in this implementation. With just 2, you can choose which audio to hear but cannot choose not to hear any audio at all. So, I added a button to the control menu to toggle the audio on and off. The default state is Mute. If the selected stream is your own screen or does not include an audio track, the button will automatically be disabled.

![image](https://github.com/screego/server/assets/20394825/7ab0f8a4-9d2b-475a-95b9-f573a975d3c1)
![image](https://github.com/screego/server/assets/20394825/9ee26c8b-53ee-41e3-90b9-7e3008ce7f19)
![image](https://github.com/screego/server/assets/20394825/0b23587d-628f-4bef-a153-2ca953f66512)

Of course, even in browsers that do not support audio sharing, such as Firefox, you can still receive audio.

I believe this approach maintains simplicity.

jmattheis reviewed (commented):
I still think this feature doesn't entirely fit within the project's scope. However, it can be argued that system audio is a part of screen sharing and thus enhances the experience of using screego.

strvert commented on ui/src/useRoom.ts:
```diff
@@ -144,9 +145,10 @@ const clientSession = async ({
     peer.ontrack = (event) => {
+        const kind = event.track.kind;
         const stream = new MediaStream();
         stream.addTrack(event.track);
-        onTrack(stream);
+        onTrack(stream, kind);
```
Initially, I considered implementing it that way as well. However, with the current track handler, a new MediaStream is created upon receiving, so in order to include multiple tracks in a single MediaStream, we would need to change the position where the MediaStream is created and move it to a higher scope.

Additionally, MediaStream is merely a view of the underlying WebRTC tracks, and I didn't think that having more of them would cause significant performance issues. However, I might be mistaken on this point.

I implemented it this way to keep the changes small, but if it's necessary to combine them into one, I will adjust accordingly.

strvert commented:
I expelicitly set `echoCancellation` and `noiseSuppression` to false to disable audio effects applied by the browser.
I combined the MediaStreams into one and added the AudioTrack and VideoTrack to it.
Additionally, the audio icon will not be displayed when there is no AudioTrack.

strvert commented:
For consistency with other features, I added the hotkey 'a' for audio. Along with this, an issue arose from the timing of initilizing the audio element reference, so I replaced it with the more appropriate useRef.

Mondrethos commented:
Any chance this will continue?

Mondrethos commented:
Any chance this will continue? @jmattheis @strvert

beyondgrayzone commented:
If anyone is interested, latest changes from this PR can be found in this release https://github.com/beyondgrayzone/screego-server/releases/tag/v1.11.3.

Bristopher commented:
> If anyone is interested, latest changes from this PR can be found in this release https://github.com/beyondgrayzone/screego-server/releases/tag/v1.11.3. 

Thank you!!

ajmandourah commented:
> If anyone is interested, latest changes from this PR can be found in this release https://github.com/beyondgrayzone/screego-server/releases/tag/v1.11.3.

would you mind making a docker image out of it. Thanks

beyondgrayzone commented:
> > If anyone is interested, latest changes from this PR can be found in this release https://github.com/beyondgrayzone/screego-server/releases/tag/v1.11.3.
> 
> would you mind making a docker image out of it. Thanks

https://hub.docker.com/r/beyondgrayzone/screego-server/tags :muscle:

jmattheis commented:
This is now supported with [v1.12.0](https://github.com/screego/server/releases/tag/v1.12.0).
